Description: Fix build problems on hurd-i386
 This patch allows Open MPI to build on Debian GNU/HURD.
 .
 Also, the memory:linux MCA component is disabled, since its use of POSIX API
 in malloc hooks called very early at startup causes troubles.
 .
 Also changes by Samuel Thibault <sthibault@debian.org>
 See bugs.debian.org/846965
Author: Pino Toscano <pino@debian.org>
Forwarded: partially
Last-Update: 2018-10-24

Index: openmpi-3.1.3~rc1/opal/util/stacktrace.c
===================================================================
--- openmpi-3.1.3~rc1.orig/opal/util/stacktrace.c
+++ openmpi-3.1.3~rc1/opal/util/stacktrace.c
@@ -588,8 +588,12 @@ int opal_util_register_stackhandlers (vo
 
     /* Setup the signals to catch */
     memset(&act, 0, sizeof(act));
+#ifdef SA_SIGINFO
     act.sa_sigaction = show_stackframe;
     act.sa_flags = SA_SIGINFO;
+#else
+    act.sa_handler = show_stackframe_handler;
+#endif
 #ifdef SA_ONESHOT
     act.sa_flags |= SA_ONESHOT;
 #else
Index: openmpi-3.1.3~rc1/orte/mca/odls/base/odls_base_default_fns.c
===================================================================
--- openmpi-3.1.3~rc1.orig/orte/mca/odls/base/odls_base_default_fns.c
+++ openmpi-3.1.3~rc1/orte/mca/odls/base/odls_base_default_fns.c
@@ -997,7 +997,11 @@ void orte_odls_base_default_launch_local
     orte_app_context_t *app;
     orte_proc_t *child=NULL;
     int rc=ORTE_SUCCESS;
-    char basedir[MAXPATHLEN];
+#if !defined(MAXPATHLEN) && defined(__GLIBC__)
+    char *basedir=NULL;
+#else
+  char basedir[MAXPATHLEN];
+#endif
     int j, idx;
     int total_num_local_procs = 0;
     orte_odls_launch_local_t *caddy = (orte_odls_launch_local_t*)cbdata;
@@ -1020,7 +1024,11 @@ void orte_odls_base_default_launch_local
      * bouncing around as we execute various apps, but we will always return
      * to this place as our default directory
      */
+#if !defined(MAXPATHLEN) && defined(__GLIBC__)
+    basedir = get_current_dir_name();
+#else
     getcwd(basedir, sizeof(basedir));
+#endif
 
     /* find the jobdat for this job */
     if (NULL == (jobdat = orte_get_job_data_object(job))) {
@@ -1907,7 +1915,12 @@ int orte_odls_base_default_restart_proc(
     int rc;
     orte_app_context_t *app;
     orte_job_t *jobdat;
-    char basedir[MAXPATHLEN];
+#if !defined(MAXPATHLEN) && defined(__GLIBC__)
+    char *basedir=NULL;
+    char *dir=NULL;
+#else
+  char basedir[MAXPATHLEN];
+#endif  
     char *wdir = NULL;
     orte_odls_spawn_caddy_t *cd;
     opal_event_base_t *evb;
@@ -1921,7 +1934,11 @@ int orte_odls_base_default_restart_proc(
      * bouncing around as we execute this app, but we will always return
      * to this place as our default directory
      */
+#if !defined(MAXPATHLEN) && defined(__GLIBC__)
+    basedir = get_current_dir_name();
+#else
     getcwd(basedir, sizeof(basedir));
+#endif
 
     /* find this child's jobdat */
     if (NULL == (jobdat = orte_get_job_data_object(child->name.jobid))) {
Index: openmpi-3.1.3~rc1/ompi/include/ompi_config.h
===================================================================
--- openmpi-3.1.3~rc1.orig/ompi/include/ompi_config.h
+++ openmpi-3.1.3~rc1/ompi/include/ompi_config.h
@@ -28,6 +28,10 @@
 
 #include "opal_config.h"
 
+#ifndef PATH_MAX /* Hurd */
+#define PATH_MAX 65535
+#endif
+
 #define OMPI_IDENT_STRING OPAL_IDENT_STRING
 
 /***********************************************************************
Index: openmpi-3.1.3~rc1/opal/mca/base/mca_base_var.c
===================================================================
--- openmpi-3.1.3~rc1.orig/opal/mca/base/mca_base_var.c
+++ openmpi-3.1.3~rc1/opal/mca/base/mca_base_var.c
@@ -51,6 +51,14 @@
 #include "opal/util/opal_environ.h"
 #include "opal/runtime/opal.h"
 
+#ifndef MAXPATHLEN /* Hurd */
+#define MAXPATHLEN 65535
+#endif
+
+#ifndef PATH_MAX  /* Hurd */
+#define PATH_MAX 65535
+#endif
+
 /*
  * local variables
  */
Index: openmpi-3.1.3~rc1/orte/include/orte_config.h
===================================================================
--- openmpi-3.1.3~rc1.orig/orte/include/orte_config.h
+++ openmpi-3.1.3~rc1/orte/include/orte_config.h
@@ -26,6 +26,13 @@
 
 #include "opal_config.h"
 
+#ifndef PATH_MAX /* Hurd */
+#define PATH_MAX 65535
+#endif
+#ifndef MAXPATHLEN
+#define MAXPATHLEN 65535
+#endif
+
 #define ORTE_IDENT_STRING OPAL_IDENT_STRING
 
 #  if OPAL_C_HAVE_VISIBILITY
Index: openmpi-3.1.3~rc1/ompi/mca/fbtl/posix/fbtl_posix_preadv.c
===================================================================
--- openmpi-3.1.3~rc1.orig/ompi/mca/fbtl/posix/fbtl_posix_preadv.c
+++ openmpi-3.1.3~rc1/ompi/mca/fbtl/posix/fbtl_posix_preadv.c
@@ -28,6 +28,10 @@
 #include "ompi/constants.h"
 #include "ompi/mca/fbtl/fbtl.h"
 
+#ifndef IOV_MAX
+#define IOV_MAX 1024
+#endif
+
 ssize_t mca_fbtl_posix_preadv (mca_io_ompio_file_t *fh )
 {
     /*int *fp = NULL;*/
Index: openmpi-3.1.3~rc1/ompi/mca/fbtl/posix/fbtl_posix_pwritev.c
===================================================================
--- openmpi-3.1.3~rc1.orig/ompi/mca/fbtl/posix/fbtl_posix_pwritev.c
+++ openmpi-3.1.3~rc1/ompi/mca/fbtl/posix/fbtl_posix_pwritev.c
@@ -30,6 +30,10 @@
 #include "ompi/constants.h"
 #include "ompi/mca/fbtl/fbtl.h"
 
+#ifndef IOV_MAX
+#define IOV_MAX 1024
+#endif
+
 ssize_t  mca_fbtl_posix_pwritev(mca_io_ompio_file_t *fh )
 {
     /*int *fp = NULL;*/
Index: openmpi-3.1.3~rc1/opal/mca/pmix/pmix2x/pmix/src/mca/base/pmix_mca_base_var.c
===================================================================
--- openmpi-3.1.3~rc1.orig/opal/mca/pmix/pmix2x/pmix/src/mca/base/pmix_mca_base_var.c
+++ openmpi-3.1.3~rc1/opal/mca/pmix/pmix2x/pmix/src/mca/base/pmix_mca_base_var.c
@@ -50,6 +50,10 @@
 #include "src/util/output.h"
 #include "src/util/pmix_environ.h"
 
+#ifndef MAXPATHLEN /* Hurd */
+#define MAXPATHLEN 1024
+#endif
+
 /*
  * local variables
  */
